File: example-clients/typescript/webhook_receiver.ts
/**
 * Receive WorldSignal WEBHOOK deliveries and verify their HMAC signature.
 * Env: WS_WEBHOOK_SECRET (must match the server), WS_PORT (default 8088), WS_MAX.
 */
import { createHmac, timingSafeEqual } from "node:crypto";
import { createServer } from "node:http";

const SECRET = process.env.WS_WEBHOOK_SECRET ?? "";
const PORT = Number(process.env.WS_PORT ?? "8088");
const MAX = Number(process.env.WS_MAX ?? "0");

function validSignature(body: Buffer, header: string | undefined): boolean {
  if (!SECRET || !header) return false;
  const expected = "sha256=" + createHmac("sha256", SECRET).update(body).digest("hex");
  const a = Buffer.from(expected);
  const b = Buffer.from(header);
  return a.length === b.length && timingSafeEqual(a, b);
}

let count = 0;
const server = createServer((req, res) => {
  const chunks: Buffer[] = [];
  req.on("data", (c: Buffer) => chunks.push(c));
  req.on("end", () => {
    const body = Buffer.concat(chunks);
    if (!validSignature(body, req.headers["x-worldsignal-signature"] as string | undefined)) {
      res.writeHead(401).end();
      console.error("[webhook] REJECTED (bad signature)");
      return;
    }
    res.writeHead(200).end();
    const d = (JSON.parse(body.toString()).data ?? {}) as { severity?: string; country?: string; title?: string };
    console.log(`[webhook] ${(d.severity ?? "?").padEnd(8)} ${d.country ?? "--"}  ${d.title ?? ""}`);
    if (MAX && ++count >= MAX) { server.close(); process.exit(0); }
  });
});
server.listen(PORT, () => console.error(`[webhook] listening on :${PORT} (verifying signatures)`));

File: example-clients/python/webhook_receiver.py
#!/usr/bin/env python3
"""Receive WorldSignal WEBHOOK deliveries and verify their HMAC signature.

Point a WEBHOOK subscription's config.url at this server. Each POST carries an
`X-WorldSignal-Signature: sha256=<hex>` header over the raw body, keyed by the
deployment's WEBHOOK_SIGNING_SECRET.

Env: WS_WEBHOOK_SECRET (must match the server), WS_PORT (default 8088), WS_MAX.
Uses only the standard library.
"""
import hashlib
import hmac
import json
import os
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer

SECRET = os.environ.get("WS_WEBHOOK_SECRET", "").encode()
PORT = int(os.environ.get("WS_PORT", "8088"))
MAX = int(os.environ.get("WS_MAX", "0"))
_count = {"n": 0}


def valid_signature(body: bytes, header: str) -> bool:
    if not SECRET or not header:
        return False
    expected = "sha256=" + hmac.new(SECRET, body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, header)


class Handler(BaseHTTPRequestHandler):
    def do_POST(self):
        body = self.rfile.read(int(self.headers.get("Content-Length", "0")))
        ok = valid_signature(body, self.headers.get("X-WorldSignal-Signature", ""))
        self.send_response(200 if ok else 401)
        self.end_headers()
        if not ok:
            print("[webhook] REJECTED (bad signature)", file=sys.stderr)
            return
        d = json.loads(body).get("data", {})
        print(f"[webhook] {d.get('severity','?'):8} {d.get('country') or '--'}  {d.get('title','')}")
        _count["n"] += 1
        if MAX and _count["n"] >= MAX:
            raise KeyboardInterrupt

    def log_message(self, *_):  # silence default access logging
        pass


def main() -> int:
    srv = HTTPServer(("0.0.0.0", PORT), Handler)
    print(f"[webhook] listening on :{PORT} (verifying signatures)", file=sys.stderr)
    try:
        srv.serve_forever()
    except KeyboardInterrupt:
        pass
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
